### PR TITLE
fix: use the correct pricing currency in cn-northwest-1 region

### LINKED ISSUE #5695: Karpenter 0.34.0 failed to get the prices of on-demand instances in cn-northwest-1
### Description

**Observed Behavior**:

The Karpenter controller failed to get the prices of on-demand instances in the cn-northwest-1 region.

```
k logs karpenter-b5cb5ff9-bbbfp
{"level":"DEBUG","time":"2024-02-20T06:02:56.490Z","logger":"controller","message":"discovered karpenter version","commit":"17d6c05","version":"v0.34.0"}
{"level":"DEBUG","time":"2024-02-20T06:02:56.580Z","logger":"controller","message":"discovered region","commit":"17d6c05","region":"cn-northwest-1"}
{"level":"DEBUG","time":"2024-02-20T06:02:56.581Z","logger":"controller","message":"discovered cluster endpoint","commit":"17d6c05","cluster-endpoint":"https://xxx.xxx.cn-northwest-1.eks.amazonaws.com.cn"}
{"level":"DEBUG","time":"2024-02-20T06:02:56.585Z","logger":"controller","message":"discovered kube dns","commit":"17d6c05","kube-dns-ip":"10.100.0.10"}
{"level":"INFO","time":"2024-02-20T06:02:56.586Z","logger":"controller","message":"webhook disabled","commit":"17d6c05"}
{"level":"INFO","time":"2024-02-20T06:02:56.586Z","logger":"controller.controller-runtime.metrics","message":"Starting metrics server","commit":"17d6c05"}
{"level":"INFO","time":"2024-02-20T06:02:56.586Z","logger":"controller.controller-runtime.metrics","message":"Serving metrics server","commit":"17d6c05","bindAddress":":8000","secure":false}
{"level":"INFO","time":"2024-02-20T06:02:56.586Z","logger":"controller","message":"starting server","commit":"17d6c05","kind":"health probe","addr":"[::]:8081"}
{"level":"INFO","time":"2024-02-20T06:02:56.687Z","logger":"controller","message":"attempting to acquire leader lease karpenter/karpenter-leader-election...","commit":"17d6c05"}
{"level":"INFO","time":"2024-02-20T06:03:13.604Z","logger":"controller","message":"successfully acquired lease karpenter/karpenter-leader-election","commit":"17d6c05"}
{"level":"INFO","time":"2024-02-20T06:03:13.604Z","logger":"controller.provisioner","message":"starting controller","commit":"17d6c05"}
{"level":"INFO","time":"2024-02-20T06:03:13.604Z","logger":"controller.disruption.queue","message":"starting controller","commit":"17d6c05"}
{"level":"INFO","time":"2024-02-20T06:03:13.604Z","logger":"controller.eviction-queue","message":"starting controller","commit":"17d6c05"}
{"level":"INFO","time":"2024-02-20T06:03:13.604Z","logger":"controller","message":"Starting EventSource","commit":"17d6c05","controller":"provisioner.trigger.pod","controllerGroup":"","controllerKind":"Pod","source":"kind source: *v1.Pod"}
{"level":"INFO","time":"2024-02-20T06:03:13.604Z","logger":"controller","message":"Starting Controller","commit":"17d6c05","controller":"provisioner.trigger.pod","controllerGroup":"","controllerKind":"Pod"}
{"level":"INFO","time":"2024-02-20T06:03:13.604Z","logger":"controller","message":"Starting workers","commit":"17d6c05","controller":"provisioner.trigger.pod","controllerGroup":"","controllerKind":"Pod","worker count":10}
{"level":"INFO","time":"2024-02-20T06:03:13.604Z","logger":"controller.disruption","message":"starting controller","commit":"17d6c05"}
{"level":"INFO","time":"2024-02-20T06:03:13.604Z","logger":"controller","message":"Starting EventSource","commit":"17d6c05","controller":"nodepool.hash","controllerGroup":"karpenter.sh","controllerKind":"NodePool","source":"kind source: *v1beta1.NodePool"}
{"level":"INFO","time":"2024-02-20T06:03:13.604Z","logger":"controller","message":"Starting Controller","commit":"17d6c05","controller":"nodepool.hash","controllerGroup":"karpenter.sh","controllerKind":"NodePool"}
{"level":"INFO","time":"2024-02-20T06:03:13.605Z","logger":"controller","message":"Starting EventSource","commit":"17d6c05","controller":"provisioner.trigger.node","controllerGroup":"","controllerKind":"Node","source":"kind source: *v1.Node"}
{"level":"INFO","time":"2024-02-20T06:03:13.605Z","logger":"controller","message":"Starting Controller","commit":"17d6c05","controller":"provisioner.trigger.node","controllerGroup":"","controllerKind":"Node"}
{"level":"INFO","time":"2024-02-20T06:03:13.605Z","logger":"controller","message":"Starting EventSource","commit":"17d6c05","controller":"state.node","controllerGroup":"","controllerKind":"Node","source":"kind source: *v1.Node"}
{"level":"INFO","time":"2024-02-20T06:03:13.605Z","logger":"controller","message":"Starting Controller","commit":"17d6c05","controller":"state.node","controllerGroup":"","controllerKind":"Node"}
{"level":"INFO","time":"2024-02-20T06:03:13.605Z","logger":"controller","message":"Starting EventSource","commit":"17d6c05","controller":"state.daemonset","controllerGroup":"apps","controllerKind":"DaemonSet","source":"kind source: *v1.DaemonSet"}
{"level":"INFO","time":"2024-02-20T06:03:13.605Z","logger":"controller","message":"Starting Controller","commit":"17d6c05","controller":"state.daemonset","controllerGroup":"apps","controllerKind":"DaemonSet"}
{"level":"INFO","time":"2024-02-20T06:03:13.605Z","logger":"controller","message":"Starting EventSource","commit":"17d6c05","controller":"state.nodepool","controllerGroup":"karpenter.sh","controllerKind":"NodePool","source":"kind source: *v1beta1.NodePool"}
{"level":"INFO","time":"2024-02-20T06:03:13.605Z","logger":"controller","message":"Starting Controller","commit":"17d6c05","controller":"state.nodepool","controllerGroup":"karpenter.sh","controllerKind":"NodePool"}
{"level":"INFO","time":"2024-02-20T06:03:13.605Z","logger":"controller","message":"Starting EventSource","commit":"17d6c05","controller":"state.pod","controllerGroup":"","controllerKind":"Pod","source":"kind source: *v1.Pod"}
{"level":"INFO","time":"2024-02-20T06:03:13.605Z","logger":"controller","message":"Starting Controller","commit":"17d6c05","controller":"state.pod","controllerGroup":"","controllerKind":"Pod"}
{"level":"INFO","time":"2024-02-20T06:03:13.605Z","logger":"controller","message":"Starting EventSource","commit":"17d6c05","controller":"state.nodeclaim","controllerGroup":"karpenter.sh","controllerKind":"NodeClaim","source":"kind source: *v1beta1.NodeClaim"}
{"level":"INFO","time":"2024-02-20T06:03:13.605Z","logger":"controller","message":"Starting Controller","commit":"17d6c05","controller":"state.nodeclaim","controllerGroup":"karpenter.sh","controllerKind":"NodeClaim"}
{"level":"INFO","time":"2024-02-20T06:03:13.605Z","logger":"controller","message":"Starting EventSource","commit":"17d6c05","controller":"metrics.nodepool","controllerGroup":"karpenter.sh","controllerKind":"NodePool","source":"kind source: *v1beta1.NodePool"}
{"level":"INFO","time":"2024-02-20T06:03:13.605Z","logger":"controller","message":"Starting Controller","commit":"17d6c05","controller":"metrics.nodepool","controllerGroup":"karpenter.sh","controllerKind":"NodePool"}
{"level":"INFO","time":"2024-02-20T06:03:13.605Z","logger":"controller","message":"Starting EventSource","commit":"17d6c05","controller":"node.termination","controllerGroup":"","controllerKind":"Node","source":"kind source: *v1.Node"}
{"level":"INFO","time":"2024-02-20T06:03:13.605Z","logger":"controller","message":"Starting Controller","commit":"17d6c05","controller":"node.termination","controllerGroup":"","controllerKind":"Node"}
{"level":"INFO","time":"2024-02-20T06:03:13.605Z","logger":"controller","message":"Starting EventSource","commit":"17d6c05","controller":"metrics.pod","controllerGroup":"","controllerKind":"Pod","source":"kind source: *v1.Pod"}
{"level":"INFO","time":"2024-02-20T06:03:13.605Z","logger":"controller","message":"Starting Controller","commit":"17d6c05","controller":"metrics.pod","controllerGroup":"","controllerKind":"Pod"}
{"level":"INFO","time":"2024-02-20T06:03:13.605Z","logger":"controller","message":"Starting EventSource","commit":"17d6c05","controller":"nodepool.counter","controllerGroup":"karpenter.sh","controllerKind":"NodePool","source":"kind source: *v1beta1.NodePool"}
{"level":"INFO","time":"2024-02-20T06:03:13.605Z","logger":"controller","message":"Starting EventSource","commit":"17d6c05","controller":"nodepool.counter","controllerGroup":"karpenter.sh","controllerKind":"NodePool","source":"kind source: *v1beta1.NodeClaim"}
{"level":"INFO","time":"2024-02-20T06:03:13.605Z","logger":"controller","message":"Starting EventSource","commit":"17d6c05","controller":"nodepool.counter","controllerGroup":"karpenter.sh","controllerKind":"NodePool","source":"kind source: *v1.Node"}
{"level":"INFO","time":"2024-02-20T06:03:13.605Z","logger":"controller","message":"Starting Controller","commit":"17d6c05","controller":"nodepool.counter","controllerGroup":"karpenter.sh","controllerKind":"NodePool"}
{"level":"INFO","time":"2024-02-20T06:03:13.605Z","logger":"controller","message":"Starting EventSource","commit":"17d6c05","controller":"nodeclaim.lifecycle","controllerGroup":"karpenter.sh","controllerKind":"NodeClaim","source":"kind source: *v1beta1.NodeClaim"}
{"level":"INFO","time":"2024-02-20T06:03:13.605Z","logger":"controller","message":"Starting EventSource","commit":"17d6c05","controller":"nodeclaim.lifecycle","controllerGroup":"karpenter.sh","controllerKind":"NodeClaim","source":"kind source: *v1.Node"}
{"level":"INFO","time":"2024-02-20T06:03:13.606Z","logger":"controller","message":"Starting Controller","commit":"17d6c05","controller":"nodeclaim.lifecycle","controllerGroup":"karpenter.sh","controllerKind":"NodeClaim"}
{"level":"INFO","time":"2024-02-20T06:03:13.606Z","logger":"controller","message":"Starting EventSource","commit":"17d6c05","controller":"nodeclaim.termination","controllerGroup":"karpenter.sh","controllerKind":"NodeClaim","source":"kind source: *v1beta1.NodeClaim"}
{"level":"INFO","time":"2024-02-20T06:03:13.606Z","logger":"controller","message":"Starting EventSource","commit":"17d6c05","controller":"nodeclaim.termination","controllerGroup":"karpenter.sh","controllerKind":"NodeClaim","source":"kind source: *v1.Node"}
{"level":"INFO","time":"2024-02-20T06:03:13.606Z","logger":"controller","message":"Starting Controller","commit":"17d6c05","controller":"nodeclaim.termination","controllerGroup":"karpenter.sh","controllerKind":"NodeClaim"}
{"level":"INFO","time":"2024-02-20T06:03:13.606Z","logger":"controller.metrics.node","message":"starting controller","commit":"17d6c05"}
{"level":"INFO","time":"2024-02-20T06:03:13.606Z","logger":"controller.nodeclaim.garbagecollection","message":"starting controller","commit":"17d6c05"}
{"level":"INFO","time":"2024-02-20T06:03:13.606Z","logger":"controller","message":"Starting EventSource","commit":"17d6c05","controller":"lease.garbagecollection","controllerGroup":"coordination.k8s.io","controllerKind":"Lease","source":"kind source: *v1.Lease"}
{"level":"INFO","time":"2024-02-20T06:03:13.606Z","logger":"controller","message":"Starting Controller","commit":"17d6c05","controller":"lease.garbagecollection","controllerGroup":"coordination.k8s.io","controllerKind":"Lease"}
{"level":"INFO","time":"2024-02-20T06:03:13.606Z","logger":"controller","message":"Starting EventSource","commit":"17d6c05","controller":"nodeclaim.consistency","controllerGroup":"karpenter.sh","controllerKind":"NodeClaim","source":"kind source: *v1beta1.NodeClaim"}
{"level":"INFO","time":"2024-02-20T06:03:13.606Z","logger":"controller","message":"Starting EventSource","commit":"17d6c05","controller":"nodeclaim.consistency","controllerGroup":"karpenter.sh","controllerKind":"NodeClaim","source":"kind source: *v1.Node"}
{"level":"INFO","time":"2024-02-20T06:03:13.606Z","logger":"controller","message":"Starting Controller","commit":"17d6c05","controller":"nodeclaim.consistency","controllerGroup":"karpenter.sh","controllerKind":"NodeClaim"}
{"level":"INFO","time":"2024-02-20T06:03:13.606Z","logger":"controller.nodeclaim.garbagecollection","message":"starting controller","commit":"17d6c05"}
{"level":"INFO","time":"2024-02-20T06:03:13.606Z","logger":"controller","message":"Starting EventSource","commit":"17d6c05","controller":"nodeclaim.disruption","controllerGroup":"karpenter.sh","controllerKind":"NodeClaim","source":"kind source: *v1beta1.NodeClaim"}
{"level":"INFO","time":"2024-02-20T06:03:13.606Z","logger":"controller","message":"Starting EventSource","commit":"17d6c05","controller":"nodeclaim.disruption","controllerGroup":"karpenter.sh","controllerKind":"NodeClaim","source":"kind source: *v1beta1.NodePool"}
{"level":"INFO","time":"2024-02-20T06:03:13.606Z","logger":"controller","message":"Starting EventSource","commit":"17d6c05","controller":"nodeclaim.disruption","controllerGroup":"karpenter.sh","controllerKind":"NodeClaim","source":"kind source: *v1.Pod"}
{"level":"INFO","time":"2024-02-20T06:03:13.606Z","logger":"controller","message":"Starting Controller","commit":"17d6c05","controller":"nodeclaim.disruption","controllerGroup":"karpenter.sh","controllerKind":"NodeClaim"}
{"level":"INFO","time":"2024-02-20T06:03:13.606Z","logger":"controller","message":"Starting EventSource","commit":"17d6c05","controller":"nodeclass","controllerGroup":"karpenter.k8s.aws","controllerKind":"EC2NodeClass","source":"kind source: *v1beta1.EC2NodeClass"}
{"level":"INFO","time":"2024-02-20T06:03:13.606Z","logger":"controller","message":"Starting EventSource","commit":"17d6c05","controller":"nodeclass","controllerGroup":"karpenter.k8s.aws","controllerKind":"EC2NodeClass","source":"kind source: *v1beta1.NodeClaim"}
{"level":"INFO","time":"2024-02-20T06:03:13.606Z","logger":"controller","message":"Starting Controller","commit":"17d6c05","controller":"nodeclass","controllerGroup":"karpenter.k8s.aws","controllerKind":"EC2NodeClass"}
{"level":"INFO","time":"2024-02-20T06:03:13.606Z","logger":"controller.pricing","message":"starting controller","commit":"17d6c05"}
{"level":"INFO","time":"2024-02-20T06:03:13.606Z","logger":"controller","message":"Starting EventSource","commit":"17d6c05","controller":"nodeclaim.tagging","controllerGroup":"karpenter.sh","controllerKind":"NodeClaim","source":"kind source: *v1beta1.NodeClaim"}
{"level":"INFO","time":"2024-02-20T06:03:13.606Z","logger":"controller","message":"Starting Controller","commit":"17d6c05","controller":"nodeclaim.tagging","controllerGroup":"karpenter.sh","controllerKind":"NodeClaim"}
{"level":"INFO","time":"2024-02-20T06:03:13.705Z","logger":"controller","message":"Starting workers","commit":"17d6c05","controller":"state.node","controllerGroup":"","controllerKind":"Node","worker count":10}
{"level":"INFO","time":"2024-02-20T06:03:13.705Z","logger":"controller","message":"Starting workers","commit":"17d6c05","controller":"nodepool.hash","controllerGroup":"karpenter.sh","controllerKind":"NodePool","worker count":10}
{"level":"INFO","time":"2024-02-20T06:03:13.705Z","logger":"controller","message":"Starting workers","commit":"17d6c05","controller":"provisioner.trigger.node","controllerGroup":"","controllerKind":"Node","worker count":10}
{"level":"INFO","time":"2024-02-20T06:03:13.711Z","logger":"controller","message":"Starting workers","commit":"17d6c05","controller":"state.daemonset","controllerGroup":"apps","controllerKind":"DaemonSet","worker count":10}
{"level":"INFO","time":"2024-02-20T06:03:13.711Z","logger":"controller","message":"Starting workers","commit":"17d6c05","controller":"lease.garbagecollection","controllerGroup":"coordination.k8s.io","controllerKind":"Lease","worker count":10}
{"level":"INFO","time":"2024-02-20T06:03:13.715Z","logger":"controller","message":"Starting workers","commit":"17d6c05","controller":"nodeclaim.lifecycle","controllerGroup":"karpenter.sh","controllerKind":"NodeClaim","worker count":1000}
{"level":"INFO","time":"2024-02-20T06:03:13.716Z","logger":"controller","message":"Starting workers","commit":"17d6c05","controller":"state.nodepool","controllerGroup":"karpenter.sh","controllerKind":"NodePool","worker count":10}
{"level":"INFO","time":"2024-02-20T06:03:13.716Z","logger":"controller","message":"Starting workers","commit":"17d6c05","controller":"state.pod","controllerGroup":"","controllerKind":"Pod","worker count":10}
{"level":"INFO","time":"2024-02-20T06:03:13.716Z","logger":"controller","message":"Starting workers","commit":"17d6c05","controller":"state.nodeclaim","controllerGroup":"karpenter.sh","controllerKind":"NodeClaim","worker count":10}
{"level":"INFO","time":"2024-02-20T06:03:13.717Z","logger":"controller","message":"Starting workers","commit":"17d6c05","controller":"metrics.nodepool","controllerGroup":"karpenter.sh","controllerKind":"NodePool","worker count":1}
{"level":"INFO","time":"2024-02-20T06:03:13.717Z","logger":"controller","message":"Starting workers","commit":"17d6c05","controller":"node.termination","controllerGroup":"","controllerKind":"Node","worker count":100}
{"level":"INFO","time":"2024-02-20T06:03:13.717Z","logger":"controller","message":"Starting workers","commit":"17d6c05","controller":"nodeclaim.termination","controllerGroup":"karpenter.sh","controllerKind":"NodeClaim","worker count":100}
{"level":"INFO","time":"2024-02-20T06:03:13.717Z","logger":"controller","message":"Starting workers","commit":"17d6c05","controller":"nodepool.counter","controllerGroup":"karpenter.sh","controllerKind":"NodePool","worker count":10}
{"level":"INFO","time":"2024-02-20T06:03:13.718Z","logger":"controller","message":"Starting workers","commit":"17d6c05","controller":"nodeclaim.tagging","controllerGroup":"karpenter.sh","controllerKind":"NodeClaim","worker count":1}
{"level":"INFO","time":"2024-02-20T06:03:13.719Z","logger":"controller","message":"Starting workers","commit":"17d6c05","controller":"metrics.pod","controllerGroup":"","controllerKind":"Pod","worker count":1}
{"level":"INFO","time":"2024-02-20T06:03:13.720Z","logger":"controller","message":"Starting workers","commit":"17d6c05","controller":"nodeclaim.consistency","controllerGroup":"karpenter.sh","controllerKind":"NodeClaim","worker count":10}
{"level":"INFO","time":"2024-02-20T06:03:13.720Z","logger":"controller","message":"Starting workers","commit":"17d6c05","controller":"nodeclaim.disruption","controllerGroup":"karpenter.sh","controllerKind":"NodeClaim","worker count":10}
{"level":"INFO","time":"2024-02-20T06:03:13.720Z","logger":"controller","message":"Starting workers","commit":"17d6c05","controller":"nodeclass","controllerGroup":"karpenter.k8s.aws","controllerKind":"EC2NodeClass","worker count":10}
{"level":"DEBUG","time":"2024-02-20T06:03:13.762Z","logger":"controller.pricing","message":"updated spot pricing with instance types and offerings","commit":"17d6c05","instance-type-count":762,"offering-count":829}
{"level":"DEBUG","time":"2024-02-20T06:03:13.793Z","logger":"controller.nodeclass","message":"discovered subnets","commit":"17d6c05","ec2nodeclass":"default","subnets":["subnet-0badb8a6e9245f6a4 (cn-northwest-1c)","subnet-041443e931e095607 (cn-northwest-1a)","subnet-0a4faecead5475b6c (cn-northwest-1b)"]}
{"level":"DEBUG","time":"2024-02-20T06:03:13.809Z","logger":"controller.disruption","message":"waiting on cluster sync","commit":"17d6c05"}
{"level":"DEBUG","time":"2024-02-20T06:03:13.871Z","logger":"controller.nodeclass","message":"discovered security groups","commit":"17d6c05","ec2nodeclass":"default","security-groups":["sg-031d1d1f77c1dd325","sg-01f0cbf4488911dd0","sg-093cadda468d0333c"]}
{"level":"DEBUG","time":"2024-02-20T06:03:13.873Z","logger":"controller.nodeclass","message":"discovered kubernetes version","commit":"17d6c05","ec2nodeclass":"default","version":"1.29"}
{"level":"DEBUG","time":"2024-02-20T06:03:13.928Z","logger":"controller","message":"hydrated launch template cache","commit":"17d6c05","tag-key":"karpenter.k8s.aws/cluster","tag-value":"xxx-dev-13","count":0}
{"level":"ERROR","time":"2024-02-20T06:03:14.002Z","logger":"controller.nodeclass","message":"discovering amis from ssm, getting ssm parameter \"/aws/service/bottlerocket/aws-k8s-1.29-nvidia/arm64/latest/image_id\", ParameterNotFound: ","commit":"17d6c05","ec2nodeclass":"default","query":"/aws/service/bottlerocket/aws-k8s-1.29-nvidia/arm64/latest/image_id"}
{"level":"ERROR","time":"2024-02-20T06:03:14.017Z","logger":"controller.nodeclass","message":"discovering amis from ssm, getting ssm parameter \"/aws/service/bottlerocket/aws-k8s-1.29-nvidia/arm64/latest/image_id\", ParameterNotFound: ","commit":"17d6c05","ec2nodeclass":"default","query":"/aws/service/bottlerocket/aws-k8s-1.29-nvidia/arm64/latest/image_id"}
{"level":"DEBUG","time":"2024-02-20T06:03:14.069Z","logger":"controller.nodeclass","message":"discovered amis","commit":"17d6c05","ec2nodeclass":"default","ids":"ami-0d1bdeec409b7bd69, ami-0fadcb605ee965289, ami-0fadcb605ee965289, ami-0d26a80b71908b194","count":4}
{"level":"ERROR","time":"2024-02-20T06:03:14.122Z","logger":"controller.pricing","message":"no on-demand pricing found","commit":"17d6c05"}
{"level":"ERROR","time":"2024-02-20T06:03:14.558Z","logger":"controller.pricing","message":"no on-demand pricing found","commit":"17d6c05"}
{"level":"ERROR","time":"2024-02-20T06:03:15.068Z","logger":"controller.pricing","message":"no on-demand pricing found","commit":"17d6c05"}
{"level":"DEBUG","time":"2024-02-20T06:03:15.260Z","logger":"controller.disruption","message":"discovered instance types","commit":"17d6c05","count":297}
{"level":"DEBUG","time":"2024-02-20T06:03:15.317Z","logger":"controller.disruption","message":"discovered offerings for instance types","commit":"17d6c05","instance-type-count":297}
{"level":"DEBUG","time":"2024-02-20T06:03:15.318Z","logger":"controller.disruption","message":"discovered zones","commit":"17d6c05","zones":["cn-northwest-1c","cn-northwest-1b","cn-northwest-1a"]}
{"level":"ERROR","time":"2024-02-20T06:03:15.537Z","logger":"controller.pricing","message":"no on-demand pricing found","commit":"17d6c05"}
{"level":"ERROR","time":"2024-02-20T06:03:15.979Z","logger":"controller.pricing","message":"no on-demand pricing found","commit":"17d6c05"}
{"level":"ERROR","time":"2024-02-20T06:03:16.443Z","logger":"controller.pricing","message":"no on-demand pricing found","commit":"17d6c05"}
{"level":"ERROR","time":"2024-02-20T06:03:16.923Z","logger":"controller.pricing","message":"no on-demand pricing found","commit":"17d6c05"}
{"level":"ERROR","time":"2024-02-20T06:03:17.437Z","logger":"controller.pricing","message":"no on-demand pricing found","commit":"17d6c05"}
{"level":"ERROR","time":"2024-02-20T06:03:18.002Z","logger":"controller.pricing","message":"no on-demand pricing found","commit":"17d6c05"}
{"level":"ERROR","time":"2024-02-20T06:03:18.715Z","logger":"controller.pricing","message":"no on-demand pricing found","commit":"17d6c05"}
{"level":"ERROR","time":"2024-02-20T06:03:19.725Z","logger":"controller.pricing","message":"no on-demand pricing found","commit":"17d6c05"}
{"level":"ERROR","time":"2024-02-20T06:03:21.178Z","logger":"controller.pricing","message":"no on-demand pricing found","commit":"17d6c05"}
{"level":"ERROR","time":"2024-02-20T06:03:23.653Z","logger":"controller.pricing","message":"no on-demand pricing found","commit":"17d6c05"}
{"level":"ERROR","time":"2024-02-20T06:03:28.210Z","logger":"controller.pricing","message":"no on-demand pricing found","commit":"17d6c05"}
{"level":"ERROR","time":"2024-02-20T06:03:36.851Z","logger":"controller.pricing","message":"no on-demand pricing found","commit":"17d6c05"}
{"level":"ERROR","time":"2024-02-20T06:03:53.706Z","logger":"controller.pricing","message":"no on-demand pricing found","commit":"17d6c05"}
{"level":"ERROR","time":"2024-02-20T06:04:26.982Z","logger":"controller.pricing","message":"no on-demand pricing found","commit":"17d6c05"}
{"level":"ERROR","time":"2024-02-20T06:05:33.030Z","logger":"controller.pricing","message":"no on-demand pricing found","commit":"17d6c05"}
{"level":"ERROR","time":"2024-02-20T06:07:44.587Z","logger":"controller.pricing","message":"no on-demand pricing found","commit":"17d6c05"}

```

**Expected Behavior**:

Get on-demand pricing successfully.

**Reproduction Steps** (Please include YAML):

Deploying Karpenter 0.34.0 to an EKS cluster in the cn-northwest-1 region.

**Versions**:
- Chart Version: 0.34.0
- Kubernetes Version (`kubectl version`):
```
Client Version: v1.28.2
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
Server Version: v1.29.0-eks-c417bb3
```

**Fix**:

I believe the root cause is the following code:

https://github.com/aws/karpenter-provider-aws/blob/d4b9494e9696cd6e7214b746aa9178ed628d8d24/pkg/providers/pricing/pricing.go#L274-L304

The `currency` variable will NOT be changed to "CNY" since my region is `cn-northwest-1`. However, the pricing API returns something like `"pricePerUnit":{"CNY":"12.4770000000"}`. L296-L298 will skip all returned prices because "USD" cannot be found. The problem could be avoided by replacing `if p.region == "cn-north-1"` with `if strings.HasPrefix(p.region, "cn-")`. I have opened PR #5694 to do this.

### PR DESCRIPTION
**Pull request by fanyang01**
Fixes #5695 

**Description**

The Karpenter controller failed to get the prices of on-demand instances in the cn-northwest-1 region and reported the following error:

```
{"level":"ERROR","time":"2024-02-20T06:03:15.537Z","logger":"controller.pricing","message":"no on-demand pricing found","commit":"17d6c05"}
{"level":"ERROR","time":"2024-02-20T06:03:15.979Z","logger":"controller.pricing","message":"no on-demand pricing found","commit":"17d6c05"}
{"level":"ERROR","time":"2024-02-20T06:03:16.443Z","logger":"controller.pricing","message":"no on-demand pricing found","commit":"17d6c05"}
{"level":"ERROR","time":"2024-02-20T06:03:16.923Z","logger":"controller.pricing","message":"no on-demand pricing found","commit":"17d6c05"}
{"level":"ERROR","time":"2024-02-20T06:03:17.437Z","logger":"controller.pricing","message":"no on-demand pricing found","commit":"17d6c05"}
{"level":"ERROR","time":"2024-02-20T06:03:18.002Z","logger":"controller.pricing","message":"no on-demand pricing found","commit":"17d6c05"}
```

The root cause is the following code:

https://github.com/aws/karpenter-provider-aws/blob/d4b9494e9696cd6e7214b746aa9178ed628d8d24/pkg/providers/pricing/pricing.go#L274-L304

The `currency` variable will NOT be changed to "CNY" since my region is `cn-northwest-1`. However, the pricing API returns something like `"pricePerUnit":{"CNY":"12.4770000000"}`. L296-L298 will skip all returned prices because "USD" cannot be found. The problem could be avoided by replacing `if p.region == "cn-north-1"` with `if strings.HasPrefix(p.region, "cn-")`.

**How was this change tested?**

**Does this change impact docs?**
- [ ] Yes, PR includes docs updates 
- [ ] Yes, issue opened: # 
- [x] No

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by fanyang01** [comment]
@liafizan I'm not sure how to write a test case for this, as I cannot find similar test cases in this repo. Could you please provide some guidance?

**Comment by liafizan** [comment]
> @liafizan I'm not sure how to write a test case for this, as I cannot find similar test cases in this repo. Could you please provide some guidance?

Please take a look at suite_test.go. This is a BDD approach and you would see some sections for actions to happen before & after every test. Then comes the actual test and your test might look like: 
```
It("should return correct amount... for CN...", func() {}
// func() is an anonymous function where you would write the test 
// Reading on the others, you can see that some of AWS functionality has been mocked and provides dummy result
// You would probably use the same approach and test if the pricing provider returns correct result when region is CN
```
Hope this helps you.

**Comment by engedaam** [comment]
@fanyang01 Can you also add a description to the PR? So that we can have an understand of the issue and goals of the PR

**Comment by fanyang01** [comment]
@liafizan Thanks for your help! I will try to add a test case based on your demonstration.

@engedaam Sure, I have added a description.

**Comment by engedaam** [comment]
@fanyang01 Any updates here?

**Comment by fanyang01** [comment]
@engedaam Sorry for the delay. I'm currently engaged in a project with a high workload. I will prioritize adding the test case within the upcoming days of this week.

File: pkg/fake/pricingapi.go [hunk 1]
```diff
@@ -51,6 +51,10 @@ func (p *PricingAPI) GetProductsPagesWithContext(_ aws.Context, _ *pricing.GetPr
 }
 
 func NewOnDemandPrice(instanceType string, price float64) aws.JSONValue {
+	return NewOnDemandPriceInCurrency(instanceType, price, "USD")
+}
+
+func NewOnDemandPriceInCurrency(instanceType string, price float64, currency string) aws.JSONValue {
```
Comment (engedaam) [inline pkg/fake/pricingapi.go:57]: Just one nit, outside of that looks good to me!

```suggestion
func NewOnDemandPriceWithCurrency(instanceType string, price float64, currency string) aws.JSONValue {
```

**Review by engedaam — COMMENTED** [review]
/karpenter snapshot

File: pkg/fake/pricingapi.go (hunk 1 above)
Comment (fanyang01) [reply pkg/fake/pricingapi.go]: Done. Thanks for your review.

**Comment by engedaam** [comment]
Seems like the test is failing. Once you get the fixed, we can merge 

```
Enter [It] should update on-demand pricing with response from the pricing API when in the CN partition - /home/runner/work/karpenter-provider-aws/karpenter-provider-aws/pkg/controllers/pricing/suite_test.go:300 @ 03/08/24 17:28:05.064
    logger.go:146: 2024-03-08T17:28:05.065Z	DEBUG	pricing/pricing.go:388	updated spot pricing with instance types and offerings	{"instance-type-count": 763, "offering-count": 1}
  [FAILED] Unexpected error:
      <*fmt.wrapError | 0xc000787b20>: 
      updating pricing, no on-demand pricing found
      {
          msg: "updating pricing, no on-demand pricing found",
          err: <*errors.errorString | 0xc000d82a70>{
              s: "no on-demand pricing found",
          },
      }
  occurred
```

**Comment by fanyang01** [comment]
It appears that the test failed because the temporary mock pricing provider was not connected to the existing pricing controller. I have updated the test to utilize a temporary pricing controller instead. Hopefully, it will work now.

File: pkg/controllers/pricing/suite_test.go [hunk 2]
```diff
@@ -297,4 +297,35 @@ var _ = Describe("Pricing", func() {
 		Expect(ok).To(BeTrue())
 		Expect(price).To(BeNumerically("==", 1.10))
 	})
+	It("should update on-demand pricing with response from the pricing API when in the CN partition", func() {
+		tmpPricingProvider := pricing.NewProvider(ctx, awsEnv.PricingAPI, awsEnv.EC2API, "cn-anywhere-1")
+		tmpController := controllerspricing.NewController(tmpPricingProvider)
```
Comment (engedaam) [inline pkg/controllers/pricing/suite_test.go:302]: Why are we creating a new controller? why can't we use the existing controller used in every other test?

Comment (fanyang01) [reply pkg/controllers/pricing/suite_test.go:302]: A pricing provider will be updated by `pricing.Controller.Reconcile()`. However, the existing `controller` is initialized with the default pricing provider (for `us-west-2`) and cannot be modified:

https://github.com/aws/karpenter-provider-aws/blob/c2016b2dcaf23ff2de6b98e169f320140520948a/pkg/controllers/pricing/suite_test.go#L63

Therefore, I have to create a new `pricing.Controller` to update the `tmpPricingProvider`.

**Review by engedaam — APPROVED** [review]
LGTM 🚀

**Comment by fanyang01** [comment]
@engedaam Thanks! I don't have permission to merge this PR. Could you please merge it?

**Pull request merged**